### PR TITLE
Add methods for writing data to temp files

### LINKED ISSUE #10: Add methods for writing data to temp files
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/tempy/issues/10)

I'd like to consolidate the functionality of https://github.com/sindresorhus/temp-write into this module. While doing that, look for ways to improve the original functionality and maybe add some more useful temp-related things.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)

### Submitted pull Requests
- [#22 Add methods for writing data to temp files](https://issuehunt.io/r/sindresorhus/tempy/pull/22)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/tempy/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by Richienb**
Fixes #10 

cc @sindresorhus

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#10: Add methods for writing data to temp files](https://issuehunt.io/repos/86448162/issues/10)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -3,9 +3,26 @@ const fs = require('fs');
 const path = require('path');
 const uniqueString = require('unique-string');
 const tempDir = require('temp-dir');
+const isStream = require('is-stream');
+const pify = require('pify');
+const writeFile = pify(fs.writeFile);
```
Comment (sindresorhus) [inline index.js]: You can just use `fs.promises`.

File: index.js [hunk 2]
```diff
@@ -3,9 +3,26 @@ const fs = require('fs');
...
 
 const getPath = () => path.join(tempDir, uniqueString());
 
+const writeStream = async (filePath, fileContent) => new Promise((resolve, reject) => {
+	const writable = fs.createWriteStream(filePath);
+
+	fileContent
```
Comment (sindresorhus) [inline index.js]: Use `stream.pipeline` instead of `.pipe`. It handles errors better. Read up on how it works.

File: index.js [hunk 3]
```diff
@@ -28,6 +45,24 @@ module.exports.directory = () => {
 	return directory;
 };
 
+module.exports.write = async (fileContent, options) => {
```
Comment (sindresorhus) [inline index.js]:
```suggestion
module.exports.write = async (data, options) => {
```

File: test.js [hunk 4]
```diff
@@ -34,6 +36,31 @@ test('.directory()', t => {
...
+	readable.push('unicorn');
+	readable.push(null);
+	const filePath = await tempy.write(readable);
+	t.is(fs.readFileSync(filePath, 'utf8'), 'unicorn');
+});
+
+test('.writeSync()', t => {
+	t.is(fs.readFileSync(tempy.writeSync('unicorn'), 'utf8'), 'unicorn');
+});
+
```
Comment (sindresorhus) [inline test.js]: There needs to b a lot more tests, especially regarding the stream handling, with tests for possible stream errors, incorrect streams being passed in, file system errors, etc. There's a lot that could go wrong with streams and we need to ensure we have good testing coverage.

File: index.d.ts [hunk 5]
```diff
@@ -61,6 +63,32 @@ declare const tempy: {
...
+
+	@example
+	```
+	import tempy = require('tempy');
+
+	await tempy.write('🦄');
+	//=> '/private/var/folders/3x/jf5977fn79jbglr7rk0tq4d00000gn/T/2f3d094aec2cb1b93bb0f4cffce5ebd6'
+	```
+	*/
+	write(fileContent: string | Buffer | NodeJS.ReadableStream, options?: tempy.Options): Promise<string>
```
Comment (sindresorhus) [inline index.d.ts]: It needs to handle the same as `fs.writeFile`, which includes `<string> | <Buffer> | <TypedArray> | <DataView>`, + our stream handling.

File: index.js (hunk 1 above)
Comment (Richienb) [reply index.js]: Fixed!

File: index.js (hunk 2 above)
Comment (Richienb) [reply index.js]: Fixed!

File: index.js (hunk 3 above)
Comment (Richienb) [reply index.js]: Added!

File: index.d.ts (hunk 5 above)
Comment (Richienb) [reply index.d.ts]: Fixed!

File: index.d.ts [hunk 6]
```diff
@@ -23,6 +25,8 @@ declare namespace tempy {
 			readonly name?: string;
 		}
 	>;
+
+	type WritableTypes = string | Buffer | TypedArray | DataView
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	type WritableType = string | Buffer | TypedArray | DataView
```

File: index.js [hunk 7]
```diff
@@ -1,11 +1,26 @@
...
 const tempDir = require('temp-dir');
+const isStream = require('is-stream');
+const {writeFile} = fs.promises;
 
 const getPath = () => path.join(tempDir, uniqueString());
 
+const writeStream = async (filePath, data) => new Promise((resolve, reject) => {
+	const writable = fs.createWriteStream(filePath);
+
+	stream.pipeline(data, writable, error => {
```
Comment (sindresorhus) [inline index.js]: You can await the pipeline.

File: index.js [hunk 8]
```diff
@@ -1,11 +1,26 @@
...
+		if (error) {
+			reject(error);
+			data.unpipe(writable);
+			writable.end();
```
Comment (sindresorhus) [inline index.js]: Are you sure it's necessary to do this unpipe and .end() manually? I would have assumed the `pipeline` does it.

File: index.d.ts (hunk 6 above)
Comment (Richienb) [reply index.d.ts]: Applied!

File: index.js (hunk 7 above)
Comment (Richienb) [reply index.js]: Not without pify.

File: index.js (hunk 8 above)
Comment (Richienb) [reply index.js]: Fixed!

File: index.js (hunk 7 above)
Comment (sindresorhus) [reply index.js]: Screenshot 2020-02-24 at 17 06 05

Comment (Richienb) [reply index.js]: Fixed!

File: test.js (hunk 4 above)
Comment (Richienb) [reply test.js]: I've added tests for stream errors.

**Comment by Richienb** [comment]
@sindresorhus Please re-review.

File: index.d.ts [hunk 9]
```diff
@@ -23,6 +24,8 @@ declare namespace tempy {
 			readonly name?: string;
 		}
 	>;
+
+	type WritableType = string | Buffer | TypedArray | DataView
```
Comment (sindresorhus) [inline index.d.ts]: This type makes no sense as it's not the writable type for the whole module, it's just for the `write()` method and not `writeSync()`. Better to just inline the type.

File: index.d.ts [hunk 10]
```diff
@@ -61,6 +64,32 @@ declare const tempy: {
 	*/
 	directory(): string;
 
+	/**
+	Write a string/buffer/stream to a random temp file.
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	Write data to a random temp file.
```

File: index.js [hunk 11]
```diff
@@ -3,9 +3,17 @@ const fs = require('fs');
...
+const isStream = require('is-stream');
+const stream = require('stream');
+const {promisify} = require('util');
+
+const pipeline = promisify(stream.pipeline);
+const {writeFile} = fs.promises;
 
 const getPath = () => path.join(tempDir, uniqueString());
 
+const writeStream = (filePath, data) => pipeline(data, fs.createWriteStream(filePath));
```
Comment (sindresorhus) [inline index.js]:
```suggestion
const writeStream = async (filePath, data) => pipeline(data, fs.createWriteStream(filePath));
```

File: index.d.ts (hunk 9 above)
Comment (Richienb) [reply index.d.ts]: Fixed!

File: readme.md [hunk 12]
```diff
@@ -54,6 +54,34 @@ Filename. Mutually exclusive with the `extension` option.
 
 Get a temporary directory path. The directory is created for you.
 
+### tempy.write(fileContent, options?)
+
+Write a string/buffer/stream to a random temp file.
```
Comment (sindresorhus) [inline readme.md]: Don't forget to keep the readme in sync with the TS file.

Comment (Richienb) [reply readme.md]: Fixed!

**Comment by sindresorhus** [comment]
Looks good now :)

**Pull request merged**